src/xy.js

```diff
@@ -1,51 +1,46 @@
-'use strict';
-
-import React from 'react';
-import ReactDOM from 'react-dom';
+import React, { useState, useRef, useEffect } from 'react';
 import PropTypes from 'prop-types';
 import themeable from 'react-themeable';
 import { autokey } from './autokey';
 import clamp from 'clamp';
 
-class XYControl extends React.Component {
-  _isMounted = false;
-
-  static propTypes = {
-    children: PropTypes.node.isRequired,
-    theme: PropTypes.object.isRequired,
-    x: PropTypes.number.isRequired,
-    y: PropTypes.number.isRequired,
-    xmax: PropTypes.number.isRequired,
-    ymax: PropTypes.number.isRequired,
-    isDark: PropTypes.string.isRequired,
-    onChange: PropTypes.func.isRequired
-  };
-
-  componentDidMount() {
-    this._isMounted = true;
-  }
-
-  componentWillUnmount() {
-    this._isMounted = false;
-  }
-
-  change(pos) {
-    if (!this._isMounted) return;
-    const rect = this.getOwnBoundingRect();
-    this.props.onChange({
-      x: clamp(pos.left, 0, rect.width) / rect.width * this.props.xmax,
-      y: clamp(pos.top, 0, rect.height) / rect.height * this.props.ymax
+function XYControl({ children, theme, x, y, xmax, ymax, isDark, onChange }) {
+  const xyControlContainer = useRef(null);
+  const [mounted, setMounted] = useState(false);
+  const [coords, setCoords] = useState({ start: {}, offset: {}, cb: null });
+  const top = Math.round(clamp((y / ymax) * 100, 0, 100));
+  const left = Math.round(clamp((x / xmax) * 100, 0, 100));
+  const themer = autokey(themeable(theme));
+
+  useEffect(() => {
+    setMounted(true);
+    return () => {
+      setMounted(false);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (coords.start.x) {
+      window.addEventListener('mousemove', drag);
+      window.addEventListener('mouseup', dragEnd);
+      window.addEventListener('touchmove', drag);
+      window.addEventListener('touchend', dragEnd);
```

Comment:
These listeners get added and removed repeatedly. Is there a way to create them once, on mount, remove them on unmount?

Reply:
Hmm I think we still want to add/remove these event handlers throughout the course of interacting with this control. We don't want `drag` executing is `dragStart` hasn't fired.